staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -17,77 +17,203 @@ limitations under the License.
 	"strings"
+	"sync"
+	"time"
 
+	"github.com/golang/glog"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
 	auditinternal "k8s.io/apiserver/pkg/apis/audit"
 	"k8s.io/apiserver/pkg/audit"
 )
 
 const (
+	// ModeBlocking indicates that the backend should process events directly,
+	// Serialize event, then write it to logs. This mode spend more time.
+	ModeBlocking = "blocking"
+	// ModeBuffered indicates that the audit backend should buffer audit events
+	// internally. If the buffer is full, audit event will be dropped, so no
+	// block event producer.
+	ModeBuffered = "buffered"
+)
+
+// AllowedModes is the modes known by this webhook.
+var AllowedModes = []string{
+	ModeBlocking,
+	ModeBuffered,
+}
+
+const (
 	// FormatLegacy saves event in 1-line text format.
 	FormatLegacy = "legacy"
 	// FormatJson saves event in structured json format.
 	FormatJson = "json"
+
+	// TODO: make this configurable
+	// Number of workers that concurrently write events.
+	workers = 5
+	// Buffer up to 1000 events.
+	defaultBufferSize = 1000
 )
 
+// The plugin name reported in error metrics.
+const pluginName = "log"
+
 // AllowedFormats are the formats known by log backend.
 var AllowedFormats = []string{
 	FormatLegacy,
 	FormatJson,
 }
 
-type backend struct {
+func NewBackend(out io.Writer, format, mode string, groupVersion schema.GroupVersion) (audit.Backend, error) {
+	switch mode {
+	case ModeBuffered:
+		return newBufferedBackend(out, format, groupVersion), nil
+	case ModeBlocking:
+		return newBlockingBackend(out, format, groupVersion), nil
+	default:
+		return nil, fmt.Errorf("log mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+}
+
+type bufferedBackend struct {
+	out          io.Writer
+	format       string
+	groupVersion schema.GroupVersion
+	buffer       chan *auditinternal.Event
+	wg           sync.WaitGroup
+}
+
+var _ audit.Backend = &bufferedBackend{}
+
+func newBufferedBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
+	return &bufferedBackend{
+		out:          out,
+		format:       format,
+		groupVersion: groupVersion,
+		buffer:       make(chan *auditinternal.Event, defaultBufferSize),
+		wg:           sync.WaitGroup{},
+	}
+}
+
+func (b *bufferedBackend) ProcessEvents(events ...*auditinternal.Event) {
+	for i, e := range events {
+		// Per the audit.Backend interface these events are reused after being
+		// sent to the Sink. Deep copy and send the copy to the queue.
+		event := e.DeepCopy()
+
+		// The following mechanism is in place to support the situation when audit
+		// events are still coming after the backend was shut down.
+		var sendErr error
+		func() {
+			// If the backend was shut down and the buffer channel was closed, an
+			// attempt to add an event to it will result in panic that we should
+			// recover from.
+			defer func() {
+				if err := recover(); err != nil {
```

Comment:
I think panicing and recoving is sort of a cop out for coordinating a shutdown. It'll make things hard if we ever want to do some sort of "API server shutting down please send the last events" hook for this (https://github.com/kubernetes/kubernetes/issues/50026)

Reply:
The mechanism referenced from webhook, If Shut down is called, no audit event can be buffered, but already buffered events can still be processed.